plugin/federation/federation.go

```diff
@@ -0,0 +1,335 @@
+package federation
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/99designs/gqlgen/codegen"
+	"github.com/99designs/gqlgen/codegen/config"
+	"github.com/99designs/gqlgen/codegen/templates"
+	"github.com/99designs/gqlgen/plugin"
+	"github.com/vektah/gqlparser"
+	"github.com/vektah/gqlparser/ast"
+	"github.com/vektah/gqlparser/formatter"
+)
+
+type federation struct {
+	SDL      string
+	Entities []*Entity
+}
+
+// New returns a federation plugin that injects
+// federated directives and types into the schema
+func New() plugin.Plugin {
+	return &federation{}
+}
+
+func (f *federation) Name() string {
+	return "federation"
+}
+
+func (f *federation) MutateConfig(cfg *config.Config) error {
+	entityFields := map[string]config.TypeMapField{}
+	for _, e := range f.Entities {
+		entityFields[e.ResolverName] = config.TypeMapField{Resolver: true}
+		for _, r := range e.Requires {
+			if cfg.Models[e.Name].Fields == nil {
+				model := cfg.Models[e.Name]
+				model.Fields = map[string]config.TypeMapField{}
+				cfg.Models[e.Name] = model
+			}
+			cfg.Models[e.Name].Fields[r.Name] = config.TypeMapField{Resolver: true}
+		}
+	}
+	builtins := config.TypeMap{
+		"_Service": {
+			Model: config.StringList{
+				"github.com/99designs/gqlgen/graphql/introspection.Service",
+			},
+		},
+		"_Any": {Model: config.StringList{"github.com/99designs/gqlgen/graphql.Map"}},
+		"Entity": {
+			Fields: entityFields,
+		},
+	}
+	for typeName, entry := range builtins {
+		if cfg.Models.Exists(typeName) {
+			return fmt.Errorf("%v already exists which must be reserved when Federation is enabled", typeName)
+		}
+		cfg.Models[typeName] = entry
+	}
+	cfg.Directives["external"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["requires"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["provides"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["key"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["extends"] = config.DirectiveConfig{SkipRuntime: true}
+
+	return nil
+}
+
+func (f *federation) InjectSources(cfg *config.Config) {
+	cfg.AdditionalSources = append(cfg.AdditionalSources, f.getSource(false))
+	f.setEntities(cfg)
+	s := "type Entity {\n"
+	for _, e := range f.Entities {
+		s += fmt.Sprintf("\t%s(%s: %s): %s!\n", e.ResolverName, e.FieldName, e.FieldTypeGQL, e.Name)
+	}
+	s += "}"
+	cfg.AdditionalSources = append(cfg.AdditionalSources, &ast.Source{Name: "entity.graphql", Input: s, BuiltIn: true})
+}
+
+func (f *federation) MutateSchema(s *ast.Schema) error {
+	// --- Set _Entity Union ---
+	union := &ast.Definition{
+		Name:        "_Entity",
+		Kind:        ast.Union,
+		Description: "A union unifies all @entity types (TODO: interfaces)",
+		Types:       []string{},
+	}
+	for _, ent := range f.Entities {
+		union.Types = append(union.Types, ent.Name)
+		s.AddPossibleType("_Entity", ent.Def)
+		// s.AddImplements(ent.Name, union) // Do we need this?
```

Comment:
I think this is what's causing the `if ent` in model gen.

https://github.com/99designs/gqlgen/pull/851/files#diff-777f78d80bba845e7f5493310d21fce6R119

Reply:
@vektah does that mean I won't need to check for the `key` directive if I add this line?

Reply:
Nevermind, I uncommented the line and it added _Entity to `implementors`. This seems to only be used in fragment-related things?